libnmstate/nm/applier.py

```diff
@@ -203,50 +204,67 @@ def _set_ifaces_admin_state(nm_client, ifaces_desired_state, con_profiles):
                 )
 
     for dev in devs_to_deactivate_beforehand:
-        device.deactivate(nm_client, dev)
+        device.deactivate(context, dev)
 
     # Do not remove devices that are marked for editing.
     for dev, _ in itertools.chain(master_ifaces_to_edit, ifaces_to_edit):
-        remove_devs_actions.pop(dev, None)
+        devs_to_delete_profile.pop(dev.get_iface(), None)
+        devs_to_delete.pop(dev.get_iface(), None)
 
     for ifname in new_master_not_enslaved_ifaces:
-        device.activate(nm_client, dev=None, connection_id=ifname)
+        device.activate(context, dev=None, connection_id=ifname)
+    context.wait_all_finish()
 
     for ifname in new_ifaces_to_activate:
-        device.activate(nm_client, dev=None, connection_id=ifname)
+        device.activate(context, dev=None, connection_id=ifname)
+    context.wait_all_finish()
 
     for dev, con_profile in master_ifaces_to_edit:
-        device.modify(nm_client, dev, con_profile)
+        device.modify(context, dev, con_profile)
+    context.wait_all_finish()
 
     for ifname in new_ovs_port_to_activate:
-        device.activate(nm_client, dev=None, connection_id=ifname)
+        device.activate(context, dev=None, connection_id=ifname)
+    context.wait_all_finish()
 
     for ifname in new_ovs_interface_to_activate:
-        device.activate(nm_client, dev=None, connection_id=ifname)
+        device.activate(context, dev=None, connection_id=ifname)
+    context.wait_all_finish()
 
     for dev, con_profile in ifaces_to_edit:
-        device.modify(nm_client, dev, con_profile)
+        device.modify(context, dev, con_profile)
+    context.wait_all_finish()
 
     for ifname in new_vlan_ifaces_to_activate:
-        device.activate(nm_client, dev=None, connection_id=ifname)
+        device.activate(context, dev=None, connection_id=ifname)
+    context.wait_all_finish()
 
     for ifname in new_vxlan_ifaces_to_activate:
-        device.activate(nm_client, dev=None, connection_id=ifname)
+        device.activate(context, dev=None, connection_id=ifname)
+    context.wait_all_finish()
 
-    for dev, actions in remove_devs_actions.items():
-        for action in actions:
-            action(nm_client, dev)
+    for dev in devs_to_delete_profile.values():
+        device.deactivate(context, dev)
+    context.wait_all_finish()
+
+    for dev in devs_to_delete_profile.values():
+        device.delete(context, dev)
+    context.wait_all_finish()
+
+    for dev in devs_to_delete.values():
+        device.delete_device(context, dev)
+    context.wait_all_finish()
```

Comment:
It is important to define what is "finish". i.e Is finish to wait until the signal of state changed? Or to wait until callback_finish is executed?

Reply:
Added a comment to `wait_all_finish()` in `nm/context.py`.

```
Block till all async actions been marked as finished via
`finish_async()` or anyone failed by `fail()`.
```